pkg/controllers/nodeclass/status/validation.go

```diff
@@ -0,0 +1,171 @@
+/*
+Portions Copyright (c) Microsoft Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/Azure/karpenter-provider-azure/pkg/apis/v1beta1"
+)
+
+type ValidationReconciler struct{}
+
+func NewValidationReconciler() *ValidationReconciler {
+	return &ValidationReconciler{}
+}
+
+const (
+	LocalDNSValidationFailed = "LocalDNSValidationFailed"
+	validationReconcilerName = "nodeclass.validation"
+	zoneRoot                 = "."
+	zoneClusterLocal         = "cluster.local"
+	fieldVnetDNSOverrides    = "vnetDNSOverrides"
+	fieldKubeDNSOverrides    = "kubeDNSOverrides"
+)
+
+var (
+	zoneRegex = regexp.MustCompile(`^(?:[A-Za-z0-9](?:[A-Za-z0-9_-]{0,61}[A-Za-z0-9])?\.)*[A-Za-z0-9](?:[A-Za-z0-9_-]{0,61}[A-Za-z0-9])?\.?$`)
+)
+
+func (r *ValidationReconciler) Reconcile(ctx context.Context, nodeClass *v1beta1.AKSNodeClass) (reconcile.Result, error) {
+	logger := log.FromContext(ctx).WithName(validationReconcilerName)
+
+	// Run LocalDNS validation if configured
+	if nodeClass.Spec.LocalDNS != nil {
+		if err := r.validateLocalDNS(logger, nodeClass.Spec.LocalDNS); err != nil {
+			nodeClass.StatusConditions().SetFalse(
+				v1beta1.ConditionTypeValidationSucceeded,
+				err.reason,
+				err.message,
+			)
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// All validations passed
+	nodeClass.StatusConditions().SetTrue(v1beta1.ConditionTypeValidationSucceeded)
+	return reconcile.Result{}, nil
+}
+
+type validationError struct {
+	reason  string
+	message string
+}
+
+func (r *ValidationReconciler) validateLocalDNS(logger logr.Logger, localDNS *v1beta1.LocalDNS) *validationError {
+	// Validate VnetDNSOverrides
+	if err := r.validateOverridesList(logger, localDNS.VnetDNSOverrides, true); err != nil {
+		return err
+	}
+
+	// Validate KubeDNSOverrides
+	if err := r.validateOverridesList(logger, localDNS.KubeDNSOverrides, false); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ValidationReconciler) validateOverridesList(logger logr.Logger, overrides []v1beta1.LocalDNSZoneOverride, isVnetDNS bool) *validationError {
```

Comment:
Anything that can (reasonably) be done in CEL should be, I think - and much here can be ...